utils/ami-cleanup/ami_cleanup.py

```diff
@@ -0,0 +1,216 @@
+
+
+def delete_ami(ec2_client, ami_id, dry_run):
+    # Deregister AMI
+
+    try:
+        if not dry_run:
+            ec2_client.deregister_image(ImageId=ami_id)
+            logger.info(f"AMI {ami_id} deregistered")
+        else:
+            logger.info(f"AMI {ami_id} would be deregistered if this wasn't a dry run")
+    except Exception as e:
+        logger.debug(f"Error while deregistering AMI {ami_id}")
+        logger.debug(e)
+
+    # Get a list of snapshots associated with the ami
+    filter_string_1 = snap_filter_1.replace("AMI_ID", ami_id)
+    filter_string_2 = snap_filter_2.replace("AMI_ID", ami_id)
+    try:
+        snapshot_list = ec2_client.describe_snapshots(
+            Filters=[{"Name": "description", "Values": [filter_string_1]}]
+        )["Snapshots"]
+    except Exception as e:
+        logger.debug(e)
+
+    if len(snapshot_list) == 0:
+        try:
+            snapshot_list = ec2_client.describe_snapshots(
+                Filters=[{"Name": "description", "Values": [filter_string_2]}]
+            )["Snapshots"]
+        except Exception as e:
+            logger.debug(e)
+
+    if len(snapshot_list) == 0:
+        logger.error(f"No snapshots found for AMI {ami_id}")
+
+    for snapshot in snapshot_list:
+        snapshot_id = snapshot["SnapshotId"]
+        try:
+            if not dry_run:
+                ec2_client.delete_snapshot(SnapshotId=snapshot_id)
+                logger.info(f"Snapshot {snapshot_id} deleted")
+            else:
+                logger.info(
+                    f"Snapshot {snapshot_id} would be deleted if this wasn't a dry run"
+                )
+        except Exception as e:
+            logger.debug(e)
+
+
+def main():
+    args = get_arguments()
+    if args.aws_profile_list is None:
+        args.aws_profile_list = []
+    main_client = get_ec2_client(args.aws_profile, args.aws_region)
+    candidate_list = get_deletion_list(main_client, args.ami_prefix, args.keep_min)
+    aws_profile_list_to_check = [args.aws_profile] + args.aws_profile_list
+
+    deletion_list = {}
+    for profile in aws_profile_list_to_check:
+        deletion_list[profile] = []
+        ec2_client = get_ec2_client(profile, args.aws_region)
+        asg_client = get_asg_client(profile, args.aws_region)
+        for ami in candidate_list:
+            if not check_ami_is_used(ec2_client, asg_client, ami["ImageId"]):
+                deletion_list[profile].append(ami["ImageId"])
+        logger.debug(f"Deletion list for profile {profile}:")
+        logger.debug(deletion_list[profile])
+
+    deletion_list_total = deletion_list[aws_profile_list_to_check[0]]
+    for profile in deletion_list:
+        deletion_list_total = common(deletion_list_total, deletion_list[profile])
+
+    logger.info(f"List of AMIs to delete: {deletion_list_total}")
+
+    for ami_id in deletion_list_total:
+        delete_ami(ec2_client, ami_id, args.dry_run)
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
this should be wrapped in a try|except to catch any unexpected exceptions thrown